gridappsd/simulation.py

```diff
@@ -1,63 +1,140 @@
+        _log.debug("Pausing simulation")
         command = dict(command="pause")
-        self.gappsd.send(simulation_input_topic(self.simulation_id), json.dumps(command))
+        self._gapps.send(simulation_input_topic(self.simulation_id), json.dumps(command))
 
     def stop(self):
-        """Stop the simulation"""
+        """ Stop the simulation"""
+        _log.debug("Stopping simulation")
         command = dict(command="stop")
-        self.gappsd.send(simulation_input_topic(self.simulation_id), json.dumps(command))
+        self._gapps.send(simulation_input_topic(self.simulation_id), json.dumps(command))
 
     def resume(self):
-        """Resume the simulation"""
+        """ Resume the simulation"""
+        _log.debug("Resuming simulation")
         command = dict(command="resume")
-        self.gappsd.send(simulation_input_topic(self.simulation_id), json.dumps(command))
+        self._gapps.send(simulation_input_topic(self.simulation_id), json.dumps(command))
+
+    def add_onmesurement_callback(self, callback, device_filter=()):
+        self.__filterable_measurement_callback_set.add((callback, device_filter))
+
+    def add_onstart_callback(self, callback):
+        self.__on_start.add(callback)
+
+    def add_oncomplete_callback(self, callback):
+        self.__on_simulation_complete_callbacks.add(callback)
+
+    def add_ontimestep_callback(self, callback):
+        self.__on_next_timestep_callbacks.add(callback)
+
+    def __on_platformlog(self, headers, message):
+        if self.simulation_id == message['processId']:
+            _log.debug("__on_platformlog: {}".format(message))
+
+        if 'command' in message:
+            _log.debug("Command was: {}".format(message))
+
+    def __on_simulation_log(self, headers, message):
+        # Handle the callbacks here
+        if 'logMessage' in message:
+            log_message = message['logMessage']
+            # if this is the last timestamp then call the finished callbacks
+            if log_message == "incrementing to {}".format(int(self._num_timesteps)):
+                for p in self.__on_simulation_complete_callbacks:
+                    p(self)
+            elif log_message.startswith("incrementing to "):
+                timestep = log_message[len("incrementing to "):]
+                for p in self.__on_next_timestep_callbacks:
+                    p(self, int(timestep))
 
-    def simulation_main_loop(self):
+    def __onmeasurement(self, headers, message):
+        """ Call the measurement callbacks
 
-        while self._is_running:
-            time.sleep(0.1)
\ No newline at end of file
+        :param headers:
+        :param message:
+        :return:
+        """
+        sim_id = message['simulation_id']
+        timestamp = message['message']['timestamp']
+        measurements = message['message']['measurements']
+        for p in self.__filterable_measurement_callback_set:
+            p[0](self, timestamp, measurements)
```

Comment:
@craig8 Again, unclear to me what should be passed to the function.

Reply:
you shouldn't be calling this function directly it is an internal function that will call your code when a measurement is published from the simulator.